### PR TITLE
Move Breeze installation of Airflow in CI image to Python 

### PR DESCRIPTION
**Pull request by potiuk**
This change replaces a lot of bash code that was installing selected
Airflow version (from pip, github or package) with Python code, while
fixing a few outstanding issues.

* Always remove Airflow code when you set "use airflow version"

When we are running compatibility checks, we should not mount
airflow from sources (not even skipping it), because if we removed a
provider, it will be imported from the mounted sources and the provider
might still be there.

Instead of mounting airflow sources we use "none" which effectively
removes sources of airflow from CI image - instead airflow is
entirely used from packages that are build during preparation of
the compatibility check (airflow from tag in git, and providers
from the PR).

* Split installation of airflow and provider into two steps
  (Following best practices for installing Airflow in a reproducible
  way (with constraints for that version) followed by installation of
  extra packages (providers in this case using either specified
  constraints or constraints from the main image by default

* Extras can be used to perform the reproducible build for Airflow.

* Better diagnostics

* Removed a lot of Bash code

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by vincbeck — APPROVED** [review]

**Comment by potiuk** [comment]
cc: @vincbeck @eladkal @amoghrajesh  @ashb - but it's one-but-last step to remove all the remainders of bash scripts we had for the CI.

The remaining part (which I did not want to touch before 2.8.0 is out) is about the docker scripts that are used while building the images.

File: scripts/in_container/install_airflow_and_providers.py [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+DIST_FOLDER = Path("/dist")
+
+
+@contextmanager
+def ci_group(group_name: str, github_actions: bool):
+    if github_actions:
+        console.print(f"::group::{group_name[:50]}[/]", markup=False)
+        console.print(group_name, markup=False)
+    else:
+        console.print(group_name, markup=False)
```
Comment (vincbeck) [inline scripts/in_container/install_airflow_and_providers.py]:
```suggestion
    if github_actions:
        console.print(f"::group::{group_name[:50]}[/]", markup=False)
    console.print(group_name, markup=False)
```

File: scripts/in_container/install_airflow_and_providers.py [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+    try:
+        yield
+    finally:
```
Comment (vincbeck) [inline scripts/in_container/install_airflow_and_providers.py:48]: I am trying to understand this code

Comment (potiuk) [reply scripts/in_container/install_airflow_and_providers.py:48]: It's just the way how to hide the output of a single command in GITHUB_ACTIONS in a "foldable" output .
anything inside the context manager will be by default hidden in smth like this:

```
> GROUP NAME
```

And when you click on the group you will be able to see defaults. You can see plenty of those in our builds :)

Comment (potiuk) [reply scripts/in_container/install_airflow_and_providers.py:48]: Say https://github.com/apache/airflow/actions/runs/7134269185/job/19428722538#step:6:239

Screenshot 2023-12-07 at 23 28 53

Every thringle on the left you see is such a group - and by clcking it, you can unfold the details.

Comment (potiuk) [reply scripts/in_container/install_airflow_and_providers.py:48]: More info on this feature of GA here: https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#grouping-log-lines

File: scripts/in_container/install_airflow_and_providers.py (hunk 1 above)
Comment (potiuk) [reply scripts/in_container/install_airflow_and_providers.py]: Yep .  obviously :)

**Review by vincbeck — APPROVED** [review]

**Comment by potiuk** [comment]
Still a few nits but I am getting really close

**Comment by potiuk** [comment]
Finally !

**Pull request merged**